Restructure Documentation (#1482)
cocotb/cocotb: documentation/source/index.rst

##################################
Welcome to Cocotb's documentation!
##################################

..
   This documentation tries to follow https://www.divio.com/blog/documentation/ (Daniele Procida)
   Other media about the same topic:
   - https://ep2018.europython.eu/media/conference/slides/get-your-documentation-right.pdf
   - https://www.youtube.com/watch?v=t4vKPhjcMZg
   - A good example: http://docs.django-cms.org/en/latest/contributing/documentation.html#contributing-documentation

   See also https://github.com/cocotb/cocotb/wiki/Howto:-Writing-Documentation

.. toctree::
   :maxdepth: 2

   introduction

.. todo::
   Move "Contributors" section to "Development & Community"

#########

> Tutorials

Reviewer: You might want to make this a `:caption:` of the TOC so it appears in the sidebar.

Author: Ah yes, so much nicer!

##################################
Welcome to Cocotb's documentation!
##################################

..
   This documentation tries to follow https://www.divio.com/blog/documentation/ (Daniele Procida)
   Other media about the same topic:
   - https://ep2018.europython.eu/media/conference/slides/get-your-documentation-right.pdf
   - https://www.youtube.com/watch?v=t4vKPhjcMZg
   - A good example: http://docs.django-cms.org/en/latest/contributing/documentation.html#contributing-documentation

   See also https://github.com/cocotb/cocotb/wiki/Howto:-Writing-Documentation

.. toctree::
   :maxdepth: 2

>    introduction

Reviewer: It's tempting to put an "installation" section here, rather than as a tutorial - that is, split the quickstart in two.

Not sure how much we want to squeeze into one PR though.

Author: Added a TODO for now.

.. todo::
   Move "Contributors" section to "Development & Community"

#########
Tutorials
#########

..
   Tutorials - lessons that take the reader by the hand through a series of steps to complete a project
   (Example: kid cooking; learning-oriented)

   - learning by doing
   - getting started
   - inspiring confidence
   - repeatability
   - immediate sense of achievement
   - concreteness, not abstraction
   - minimum necessary explanation
   - no distractions

.. toctree::
   :maxdepth: 1
   :caption: Tutorials

   quickstart
   endian_swapper
   ping_tun_tap
   hal_cosimulation
   examples

#############
How-to Guides

> #############

Reviewer: ```suggestion
```

These are auto-inserted by `:caption:`

Author: Indeed; removed.

.. toctree::
   :maxdepth: 1
   :caption: Key topics

   troubleshooting

.. todo::
   - Add some info from :doc:`coroutines`
   - Add GPI section
   - Explain ReadOnly/ReadWrite/... phases
   - Add pitfall from https://github.com/cocotb/cocotb/issues/526#issuecomment-300371629 to troubleshooting

#########
Reference
#########

..
   Reference - technical descriptions of the machinery and its operation
   (Example: Wikipedia pages of ingredients; information-oriented)

   - structure
   - consistency
   - description
   - accuracy

.. toctree::
   :maxdepth: 1
   :caption: Reference

   building
   Python Code Library Reference <library_reference>
   C/C++ Code Library Reference <library_reference_c>
   simulator_support
   extensions

.. todo::
   - *Maybe* add a glossary (Coroutine, Driver, Monitor, Scoreboard, HDL, RTL, GPI, V(H)PI, FLI, VIP, UVM, MDV, DUT/DUV)

#######################
Development & Community
#######################

.. toctree::
   :maxdepth: 1
   :caption: Development & Community

   roadmap

.. todo::
   Add "Join us online" and "Contributing"

#############
Release Notes
#############

.. toctree::
   :maxdepth: 1
   :caption: Release Notes

>    release_notes

Reviewer: This looks a bit weird in the sidebar:

![image](https://user-images.githubusercontent.com/425260/76152426-618af600-60b7-11ea-95ac-ca17db2299bf.png)

Can we combine this under another heading?

Author: Let's just have it under "Development & Community", after all, it's the description of our development so far.

##################################
Welcome to cocotb's documentation!
##################################

..
   This documentation tries to follow https://www.divio.com/blog/documentation/ (Daniele Procida)
   Other media about the same topic:
   - https://ep2018.europython.eu/media/conference/slides/get-your-documentation-right.pdf
   - https://www.youtube.com/watch?v=t4vKPhjcMZg
   - A good example: http://docs.django-cms.org/en/latest/contributing/documentation.html#contributing-documentation

   See also https://github.com/cocotb/cocotb/wiki/Howto:-Writing-Documentation

***************
What is cocotb?
***************

**cocotb** is a *COroutine* based *COsimulation* *TestBench* environment for verifying VHDL/Verilog RTL using `Python <https://www.python.org>`_.

cocotb is completely free, open source (under the `BSD License <https://en.wikipedia.org/wiki/BSD_licenses#3-clause_license_(%22BSD_License_2.0%22,_%22Revised_BSD_License%22,_%22New_BSD_License%22,_or_%22Modified_BSD_License%22)>`_) and hosted on `GitHub <https://github.com/cocotb/cocotb>`_.

cocotb requires a simulator to simulate the HDL design

> and has been used with a variety of simulators on Linux, Windows and Mac OS.

Reviewer 2: ```suggestion
and has been used with a variety of simulators on Linux, Windows and macOS.
```

Please check the :ref:`simulator-support` page for specifics.

> A (possibly older) version of cocotb can be used live in a web-browser using `EDA Playground <https://www.edaplayground.com>`_.

Reviewer 2: ```suggestion
A (possibly older) version of cocotb can be used live in a web browser on `EDA Playground <https://www.edaplayground.com>`_.
```

************************
How is cocotb different?
************************

> cocotb encourages the same philosophy of design re-use and randomized testing as UVM, however is implemented in Python rather than SystemVerilog.

Reviewer 2: ```suggestion
cocotb encourages the same philosophy of design re-use and randomized testing as `UVM <https://en.wikipedia.org/wiki/Universal_Verification_Methodology>`_, however is implemented in Python rather than SystemVerilog.
```

With cocotb, VHDL/Verilog/SystemVerilog are normally only used for the design itself, not the testbench.

> cocotb has built-in support for integrating with the `Jenkins <https://jenkins.io/>`_ continuous integration system.

Reviewer 2: ```suggestion
cocotb has built-in support for integrating with continuous integration systems, such as Jenkins, GitLab, etc. through standardized, machine-readable test reporting formats.
```

cocotb was specifically designed to lower the overhead of creating a test.

cocotb automatically discovers tests so that no additional step is required to add a test to a regression.

All verification is done using Python which has various advantages over using SystemVerilog or VHDL for verification:

> * Writing Python is **fast** - it's a very productive language

Reviewer 2: ```suggestion
* Writing Python is **fast** - it's a very productive language.
```

> * It's **easy** to interface to other languages from Python

Reviewer 2: ```suggestion
* It's **easy** to interface to other languages from Python.
```

> * Python has a huge library of existing code to **re-use** like `packet generation <https://www.secdev.org/projects/scapy/>`_ libraries.

Reviewer 2: ```suggestion
* Python has a huge library of existing code to **re-use**.
```

(scapy is great, but slightly overspecific for such an introduction.)

The Design Under Test (DUT) is instantiated as the toplevel in the simulator without any wrapper code.
cocotb drives stimulus onto the inputs to the DUT (or further down the hierarchy) and monitors the outputs directly from Python.

.. image:: diagrams/svg/cocotb_overview.svg

A test is simply a Python function.
At any given time either the simulator is advancing time or the Python code is executing.
The :keyword:`await` keyword is used to indicate when to pass control of execution back to the simulator.
A test can spawn multiple coroutines, allowing for independent flows of execution.

************
Contributors
************

.. spelling::
   McGregor
   Grimwood

cocotb was developed by `Potential Ventures <https://potential.ventures>`_ with the support of
`Solarflare Communications Ltd <https://www.solarflare.com/>`_
and contributions from Gordon McGregor and Finn Grimwood
(see `contributors <https://github.com/cocotb/cocotb/graphs/contributors>`_ for the full list of contributions).

We also have a list of talks and papers, libraries and examples at our Wiki page
`Further Resources <https://github.com/cocotb/cocotb/wiki/Further-Resources>`_.

> Feel free to add links to cocotb-related content that we are still missing!

Reviewer 2: Please remove the whole contributors section for now, it's outdated. You can replace it with something like:

```rst
cocotb is developed and maintained by an active community under the umbrella of the `Free and Open Source Silicon Foundation <https://www.fossi-foundation.org>`_. 

Our GitHub repository contains a list of all `contributors <https://github.com/cocotb/cocotb/graphs/contributors>`_. 
```

Reviewer: I'm not super keen on doing this as part of what is essentially just a move.

I'd rather we revert this for now, and do a separate PR to remove attribution we feel unecessary. Hiding an attribution removal amongst lots of moves feels dishonest (and I plan to squash the commits).

Author: Reverted now, we can do it together with the small copyright notices in the HTML footer in conf.py in another PR.

.. toctree::
   :maxdepth: 1
   :caption: Tutorials

>    :name: tutorials

Reviewer: What's the purpose of this?

Author: It does link to this index page. The pages in the toctree are separate pages, so without ``:name:``, there wouldn't be a way to link to e.g. "all tutorials".
